src/main/java/com/arpnetworking/metrics/mad/parsers/GraphitePlaintextToRecordParser.java

```
@@ -0,0 +1,262 @@
+     *
+     * @param datagram a datagram
+     * @return A list of {@link DefaultRecord.Builder}
+     * @throws ParsingException if the datagram is not parsable as graphite plaintext formatted message
+     */
+    public List<Record> parse(final ByteBuffer datagram) throws ParsingException {
+        // CHECKSTYLE.OFF: IllegalInstantiation - This is the recommended way
+        final String datagramAsString = new String(datagram.array(), Charsets.UTF_8);
+        final ImmutableList.Builder<Record> recordListBuilder = ImmutableList.builder();
+        final DateTime now = DateTime.now();
+        for (final String line : LINE_SPLITTER.split(datagramAsString)) {
+            // CHECKSTYLE.ON: IllegalInstantiation
+            final Matcher matcher = GRAPHITE_PATTERN.matcher(line);
+            if (!matcher.matches()) {
+                throw new ParsingException("Invalid graphite line", line.getBytes(Charsets.UTF_8));
+            }
+
+            // Annotations
+            final Map<String, String> dimensions = Maps.newHashMap();
+
+            // Add global dimensions (lowest priority)
+            // IMPORTANT: Subsequently processed tags can overwrite these
+            dimensions.putAll(_globalTags);
+
+            // Parse the name
+            String name = parseName(datagram, matcher.group("NAME"));
+            if (_parseCarbonTags) {
+                // IMPORTANT: Subsequently processed tags can overwrite these
+                final String[] parts = name.split(";");
+                name = parts[0];
+                for (int i = 1; i < parts.length; ++i) {
+                    final String[] subparts = parts[i].split("=");
+                    if (subparts.length == 2 && !subparts[0].isEmpty() && !subparts[1].isEmpty()) {
+                        dimensions.put(subparts[0], subparts[1]);
+                    }
+                    // TODO(ville): should we rate limit log that we potentially dropped something?
```

I don't think we should fail the whole bytebuffer for invalid data.  If we want to keep the parser processing multiple records, I think it should discard invalid lines, but still process valid lines.

---

I agree. However, this is a large discussion imo and affects more sinks than just this one. I would like to discuss with you and make changes separately. The goal here was to raise the issue while achieving parity with the current design.